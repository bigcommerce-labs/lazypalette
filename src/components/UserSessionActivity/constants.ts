// tslint:disable:max-line-length
export const enum Messages {
    LogoutAlertHeading = 'You have been logged out.',
    LogoutAlertBody = 'For security reasons you have been automatically logged out. Please log in again to access Store Design.',
    Ok = 'Ok',
}
// tslint:enable:max-line-length
