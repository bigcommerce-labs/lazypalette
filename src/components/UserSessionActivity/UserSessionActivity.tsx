import React, { Component } from 'react';
import PageVisibility from 'react-page-visibility';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';

import { sessionHeartbeatResponse } from '../../actions/sessionHeartbeat';
import BrowserContext, { Browser } from '../../context/BrowserContext';
import { State } from '../../reducers/reducers';
import { throttledHeartbeat } from '../../services/sessionHeartbeat/sessionHeartbeat';

import AlertModal from '../Modal/AlertModal/AlertModal';

import { Messages, SessionLinks } from './constants';

interface UserSessionActivityProps {
    heartbeatResponse: Dispatch<State>;
    oauthBaseUrl: string;
    children: JSX.Element;
    isLoggedIn: boolean;
    queryParams?: string;
}

/*
 * HoC which handles keeping the users session active. When used, any events which this component listens for will
 * be sent to a throttled heartbeat which will keep the users session active in the control panel.
 */
export class UserSessionActivity extends Component<UserSessionActivityProps> {
    componentDidMount() {
        this.handleHeartbeat();
    }

    handleChange = (isVisible: boolean) => {
        if (isVisible) {
            this.handleHeartbeat();
        }
    };

    handleHeartbeat = () => {
        const { heartbeatResponse, oauthBaseUrl } = this.props;

        throttledHeartbeat(oauthBaseUrl, heartbeatResponse);
    };

    handleClose = (_window: Window) => {
        const { oauthBaseUrl, queryParams = '' } = this.props;

        _window.location.assign(`${oauthBaseUrl}${SessionLinks.StoreDesign}${queryParams}`);
    };

    render() {
        const { children, isLoggedIn } = this.props;

        return (
            <PageVisibility onChange={this.handleChange}>
                <div onClick={this.handleHeartbeat}>
                    {!isLoggedIn &&
                        <BrowserContext.Consumer>
                            {({ _window }: Browser) =>
                                <AlertModal
                                    title={Messages.LogoutAlertHeading}
                                    body={Messages.LogoutAlertBody}
                                    primaryAction={() => this.handleClose(_window)}
                                />
                            }
                        </BrowserContext.Consumer>
                    }
                    {children}
                </div>
            </PageVisibility>
        );
    }
}

const mapDispatchToProps = {
    heartbeatResponse: sessionHeartbeatResponse,
};

const mapStateToProps = (state: State) => ({
    isLoggedIn: state.session.isLoggedIn,
});

export default connect(mapStateToProps, mapDispatchToProps)(UserSessionActivity);
