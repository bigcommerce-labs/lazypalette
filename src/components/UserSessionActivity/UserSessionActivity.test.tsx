import { shallow } from 'enzyme';
import React from 'react';

import { UserSessionActivity } from './UserSessionActivity';

jest.mock('../../services/sessionHeartbeat');

describe('UserSessionActivity', () => {
    const oauthBaseUrl = 'https://login.service.bcdev';

    it('renders consistently', () => {
        const mockCallBack = jest.fn();
        const component = shallow(
            <UserSessionActivity oauthBaseUrl={oauthBaseUrl} heartbeatResponse={mockCallBack}>
                <p>
                    Ok!
                </p>
            </UserSessionActivity>
        );
        expect(component).toMatchSnapshot();
    });

    describe('pageActivityHandler', () => {
        describe('onClick', () => {
            it('triggers the sessionHeartbeat', () => {
                const mockHandler = jest.fn();
                const component = shallow(
                    <UserSessionActivity oauthBaseUrl={oauthBaseUrl} heartbeatResponse={mockHandler}>
                        <p>meow meow</p>
                    </UserSessionActivity>
                );

                const div = component.find('div');
                div.simulate('click');
                expect(mockHandler.mock.calls.length).toEqual(1);
            });
        });
    });
});
