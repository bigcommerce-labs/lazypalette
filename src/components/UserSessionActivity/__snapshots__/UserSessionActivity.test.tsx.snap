// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserSessionActivity renders consistently 1`] = `
<div>
  <p>
    Ok!
  </p>
</div>
`;

exports[`UserSessionActivity when the user is logged out renders an AlertModal 1`] = `
<Component
  heartbeatResponse={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
    }
  }
  isLoggedIn={false}
  oauthBaseUrl="https://login.service.bcdev"
>
  <div>
    <AlertModal
      actionText="Ok"
      body="For security reasons you have been automatically logged out. Please log in again to access Store Design."
      primaryAction={[Function]}
      title="You have been logged out."
    >
      <Modal
        isTransparent={false}
        title="You have been logged out."
      >
        <styled.div
          onClick={[Function]}
          theme={
            Object {
              "layers": Object {
                "higher": 100,
              },
            }
          }
        >
          <div
            className="sc-bxivhb boVVrw"
            onClick={[Function]}
          >
            <styled.div
              isTransparent={false}
              theme={
                Object {
                  "layers": Object {
                    "higher": 100,
                  },
                }
              }
            >
              <div
                className="sc-ifAKCX glQnqD"
              />
            </styled.div>
            <styled.div
              onClick={[Function]}
              theme={
                Object {
                  "colors": Object {
                    "empty": "#FFFFFF",
                    "guideText": "#ABB1BE",
                  },
                  "elevation": Object {
                    "floating": "0 2px 12px rgba(48, 53, 64, 0.2)",
                  },
                  "layers": Object {
                    "highest": 500,
                  },
                }
              }
            >
              <div
                className="sc-htoDjs dirVeL"
                onClick={[Function]}
              >
                <styled.div>
                  <div
                    className="sc-bZQynM ceVmct"
                  >
                    <styled.h3
                      theme={
                        Object {
                          "colors": Object {
                            "primaryText": "#303540",
                          },
                          "typography": Object {
                            "fontSize": Object {
                              "larger": "24px",
                            },
                            "fontWeight": Object {
                              "normal": 600,
                            },
                          },
                        }
                      }
                    >
                      <h3
                        className="sc-gzVnrw idAWTr"
                      >
                        You have been logged out.
                      </h3>
                    </styled.h3>
                    <styled.div
                      onClick={[Function]}
                      theme={
                        Object {
                          "colors": Object {
                            "background": "#000000",
                            "primaryText": "#AAAAAA",
                          },
                        }
                      }
                    >
                      <div
                        className="sc-htpNat gAQftg"
                        onClick={[Function]}
                      />
                    </styled.div>
                  </div>
                </styled.div>
                <styled.div
                  theme={
                    Object {
                      "colors": Object {
                        "primaryText": "#AAAAAA",
                      },
                      "typography": Object {
                        "fontSize": Object {
                          "small": "16px",
                        },
                      },
                    }
                  }
                >
                  <div
                    className="sc-EHOje cnVQSR"
                  >
                    <styled.div>
                      <div
                        className="sc-VigVT iwIlAC"
                      >
                        <styled.div>
                          <div
                            className="sc-iwsKbI iuVdIU"
                          >
                            For security reasons you have been automatically logged out. Please log in again to access Store Design.
                          </div>
                        </styled.div>
                        <styled.div>
                          <div
                            className="sc-jTzLTM nkRIC"
                          >
                            <ButtonInput
                              classType="primary"
                              onClick={[Function]}
                              testId="ok"
                              type="button"
                            >
                              <styled.div>
                                <div
                                  className="sc-bdVaJa cubKnN"
                                >
                                  <styled.button
                                    border={true}
                                    classType="primary"
                                    data-test-id="ok"
                                    onClick={[Function]}
                                    theme={
                                      Object {
                                        "colors": Object {
                                          "background": "#F5F7FA",
                                          "brandPrimary": "#273A8A",
                                          "primary": "#3F69FE",
                                        },
                                        "typography": Object {
                                          "fontFamily": Object {
                                            "sans": "\\"Source Sans Pro\\", \\"Helvetica Neue\\", Arial, sans-serif",
                                          },
                                        },
                                      }
                                    }
                                    type="button"
                                  >
                                    <button
                                      className="sc-bwzfXH kBSfBk"
                                      data-test-id="ok"
                                      onClick={[Function]}
                                      type="button"
                                    >
                                      Ok
                                    </button>
                                  </styled.button>
                                </div>
                              </styled.div>
                            </ButtonInput>
                          </div>
                        </styled.div>
                      </div>
                    </styled.div>
                  </div>
                </styled.div>
              </div>
            </styled.div>
          </div>
        </styled.div>
      </Modal>
    </AlertModal>
    <p>
      meow meow
    </p>
  </div>
</Component>
`;
